### PR TITLE
 feat(rs-sdk): fetch document, identity, data contract and list documents

### PR DESCRIPTION
**Pull request by lklimek**
## Issue being fixed or feature implemented

`rs-sdk` is a high-level crate that provides external developers with easy API interface to access the Dash Platform.

Identity, Data Contract and Document objects are base elements of rs-sdk. In this PR, we create initial version of rs-sdk.

## What was done?

1. fetch Document, Data Contract and Identity using Fetch trait
2. list Documents using List trait
3. mocking of all above
4. unit tests

Notes for reviewer: 

1. changes in dapi-client already reviewed in #1417 

## How Has This Been Tested?

* added tests to rs-sdk/tests/mock_*.rs
* added tests that run against running platform (need updating some CONSTs before run)

## Breaking Changes

None

## Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have made corresponding changes to the documentation

**For repository code-owners and collaborators only**
- [ ] I have assigned this pull request to a milestone

### CODE REVIEW

**Review by fominok — APPROVED** [review]

File: packages/rs-sdk/src/platform/query.rs [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    fn query(self) -> Result<GetDataContractRequest, Error> {
+        let id = self.to_vec();
+        Ok(GetDataContractRequest { id, prove: true })
+    }
+}
+
+impl Query<GetIdentityRequest> for dpp::prelude::Identifier {
+    fn query(self) -> Result<GetIdentityRequest, Error> {
+        let id = self.to_vec();
+        Ok(GetIdentityRequest { id, prove: true })
```
[the remark this answers is not in the document]
Comment (lklimek) [reply packages/rs-sdk/src/platform/query.rs:88]: Updated the trait to add support in future in 6d08229a6c898d24bea834e604181aa9b5445d3e .

File: packages/rs-sdk/src/platform/document_query.rs [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+/// Request that is used to query documents from the Dash Platform.
+///
+/// This is an abstraction layer built on top of [GetDocumentsRequest] to address issues with missing details
+/// required to correctly verify proofs returned by the Dash Platform.
+///
+/// Conversions are implemented between this type, [GetDocumentsRequest] and [DriveQuery] using [TryFrom] trait.
+#[derive(Debug, Clone, serde::Serialize)]
+pub struct DocumentQuery {
+    /// Data contract ID
+    pub data_contract: Arc<DataContract>,
```
Comment (QuantumExplorer) [inline packages/rs-sdk/src/platform/document_query.rs:35]: Don't use Arc, we're on the client, so cloning is fine. However I would do this fix in another PR and not this one. As the other PR we could get a good feel about what is being changed.

File: packages/rs-sdk/src/platform/document_query.rs [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+    /// Document type for the data contract
+    pub document_type_name: String,
+    /// `where` clauses for the query
+    pub where_clauses: Vec<WhereClause>,
+    /// `order_by` clauses for the query
+    pub order_by_clauses: Vec<OrderClause>,
+    /// queryset limit
+    pub limit: u32,
+    /// first object to start with
+    pub start: Option<Start>,
```
Comment (QuantumExplorer) [inline packages/rs-sdk/src/platform/document_query.rs:45]: I think you need a start_included bool.

File: packages/rs-sdk/src/platform/fetch.rs [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    ///
+    /// Any errors encountered during the execution are returned as [Error] instances.
+    async fn fetch<Q: Query<<Self as Fetch>::Request>>(
+        sdk: &mut Sdk,
+        query: Q,
+    ) -> Result<Option<Self>, Error> {
+        let request = query.query(sdk.prove())?;
+
+        let response = request
+            .clone()
```
Comment (QuantumExplorer) [inline packages/rs-sdk/src/platform/fetch.rs:79]: who do we need to clone the request?

Comment (QuantumExplorer) [reply packages/rs-sdk/src/platform/fetch.rs:79]: (not saying that this is a problem, just curious)

File: packages/rs-sdk/src/platform/list.rs [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+use crate::{
+    error::Error,
+    mock::MockResponse,
+    platform::{document_query::DocumentQuery, query::Query},
+    Sdk,
+};
+
+/// Trait implemented by objects that can be listed or searched.
+#[async_trait::async_trait]
+pub trait List
```
Comment (QuantumExplorer) [inline packages/rs-sdk/src/platform/list.rs:22]: Rename to FetchList or FetchMany (I think FetchMany is better but would be okay with FetchList)

**Review by QuantumExplorer — APPROVED** [review]
Very nice work, I do think you might have made your life a little more complicated than it needed to be, but this should work very nicely, and might be the right solution. I think time will tell once we start using it if it flows properly and people like it.

File: packages/rs-sdk/src/platform/document_query.rs (hunk 2 above)
Comment (QuantumExplorer) [reply packages/rs-sdk/src/platform/document_query.rs:35]: Let's leave ARC for now, we can think on how to better do this later.

File: packages/rs-sdk/src/platform/fetch.rs (hunk 4 above)
Comment (lklimek) [reply packages/rs-sdk/src/platform/fetch.rs:75]: `execute()` consumes the request, so we need to clone it here in order to log it later and pass to sdk.parse_proof.
